### PR TITLE
add FB and YT surrogates

### PR DESCRIPTION
**Pull request by ladamski**
Add FB and YT surrogates

### CODE REVIEW

File: mapping.json
Comment (kdzwinel) [inline mapping.json:39]: sorry, I missed this discussion - why would putting those regexes into "url" not work? Or it would work but we are creating `regexUrl` so that we can do the regex-specific validation in `test-mapping.js`?

File: mapping.json
Comment (kdzwinel) [inline mapping.json:3]: +1, that's a better format, no idea why I had arrays here before, that's never a good idea

File: mapping.json
Comment (kdzwinel) [inline mapping.json:43]: again, missing some details from TD discussions - do we need a separate action for each CTL target? This means that whenever we add a new target (instagram, twitter, etc.) we will have to update client apps to understand this new action.

File: mapping.json
Comment (kdzwinel) [inline mapping.json:46]: `(www\\.)?` - do we need this part if it's optional?

File: mapping.json
Comment (kdzwinel) [inline mapping.json:39]: aren't those going to match against notfacebook.com? I guess the domain check that happens first will prevent it?

**Review by kdzwinel — COMMENTED** [review]
I left some questions, but it LGTM, nicely done.

I have NOT reviewed the actual surrogates, because my assumption is that they have been reviewed in FB CTL and YT CTL projects. LMK if I should review them.

File: mapping.json
Comment (ladamski) [reply mapping.json:39]: We discussed this around here https://app.asana.com/0/0/1202673425163833/1202706967057780/f, but the gist of it is that when building `mapping.js`, the URL is escaped to build a valid regex.  If its already a regex, it would end up being double-escaped.  To avoid that (and some hacky "is it regex or not" auto-detection) we decided on having a separate property to denote a regex URL.

File: mapping.json
Comment (ladamski) [reply mapping.json:43]: This was discussed in https://app.asana.com/0/0/1201397262498310/1201415252824341/f https://app.asana.com/0/0/1201397262498310/1201415252824341/f

Basically these are per-client implementations anyway (each needs bespoke support whether it be new content scripts, CBR rules, etc).

The intent is to update these clients to ignore rules with unrecognized actions, and then trigger the corresponding code path for CTL once we port over the current implementations to rely on the TDS rules instead of their own.

File: mapping.json
Comment (ladamski) [reply mapping.json:46]: This is in the current YT surrogate URL (in the extension) so I just copied it over verbatim.

File: mapping.json
Comment (ladamski) [reply mapping.json:39]: I think we'd have that problem across the board?  Since the other URLs above are also turned into the same style regex.

Comment (kdzwinel) [reply mapping.json:39]: Gotcha, that makes sense to me although I wonder why just not convert all `url`s to `regexUrl`? Do we need `url` if we have `regexUrl`?

File: mapping.json
Comment (kdzwinel) [reply mapping.json:43]: +1 to a new action for those, but I wonder why those actions have to be youtube/facebook specific? Ideally all configuration related to new CTL targets would be separate from client implementations allowing us to rapidly add more targets. You mention:

- content scripts - but those are the surrogates, right? Those can be updated independently from the browser (remotely / as a dependency)
- CBR rules - those should be generated based on blocklist which is also remote / dependency

Asking differently - if we add Instagram CTL support tomorrow, what will have to happen on clients?

File: mapping.json
Comment (kdzwinel) [reply mapping.json:46]: cc @kzar - does this regex prefix do anything or can we drop it?

File: mapping.json
Comment (kdzwinel) [reply mapping.json:39]: hm, I think we do an initial check against "facebook.net" so I think we should be good here (cc @jdorweiler )

File: mapping.json
Comment (kzar) [reply mapping.json:46]: Good point @kdzwinel, it looks like the check (`url.match(surrogate.rule))`) does not anchor to the start of the URL anyway, so I agree that we can drop the `(www\\.)?` prefix. [This is where the check takes place](https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/develop/shared/js/background/tracker-utils.js#L84).

File: surrogates/fb-sdk.js
Comment (kzar) [inline surrogates/fb-sdk.js:80]: Just a note to say I've got [a tiny open PR to improve this script](https://github.com/duckduckgo/duckduckgo-privacy-extension/pull/1369). It might be worth waiting for that to land and including that change here, before landing this?

File: mapping.json
Comment (ladamski) [reply mapping.json:43]: Yes because each CTL features has needed some bespoke client support (native code) to support it.  We could also use a generic "clicktoload" action that covers features that don't need extra work, right?  TBH I feel like we're rehashing the TD here.

Comment (kdzwinel) [reply mapping.json:43]: "needed some bespoke client support" - I'm curious what is it, but you are totally right, this is not a PR but a TD discussion so I'm bringing it back there.

File: mapping.json
Comment (ladamski) [reply mapping.json:39]: @kdzwinel sure, we can just replace all of the `url` properties with `regexUrl` instead in `mapping.js`

File: mapping.json
Comment (kzar) [reply mapping.json:3]: Oh, let's add an optional key for click to load entity. E.g. `"clickToLoadEntity": "Facebook, Inc."`. That would help us when we come to implement Chrome MV3 Click to Load support.

**Review by kdzwinel — COMMENTED** [review]
I found couple of nits, but other than that it LGTM 👍 

I still don't understand why we need separate actions for each CTL integration, but I'll ask via asana.

File: mapping.json
Comment (ladamski) [reply mapping.json]: I updated it to `regexRule` (we deprecated the plaintext one).

Regarding  `clickToLoadEntity`, is this just for the mapping.json?  I'd rather handle Mv3 specific stuff in a separate TD and followup PR, since it seems like that'll be on my plate next anyway.

File: surrogates/fb-sdk.js
Comment (ladamski) [reply surrogates/fb-sdk.js:3]: 👍

File: scripts/test-mapping.js
Comment (ladamski) [reply scripts/test-mapping.js]: Good catch, done

File: surrogates/fb-sdk.js
Comment (ladamski) [reply surrogates/fb-sdk.js]: 👍

File: mapping.json
Comment (kzar) [reply mapping.json]: > I updated it to regexRule...

Ace, thanks for renaming to `regexRule`.

> (we deprecated the plaintext one)

How come you deprecated plantext rules? (I don't feel too strongly, but I wonder why we wouldn't use plaintext matching where that sufficed.)

> Regarding clickToLoadEntity, is this just for the mapping.json? I'd rather handle Mv3 specific stuff in a separate TD and followup PR, since it seems like that'll be on my plate next anyway.

Yea, that would just be for `mapping.json`. OK, fair enough if you'd like to do that as a follow-up. It's just if you made that change now it would make my life slightly easier. Well, if we at least agree to the name now that's enough I suppose. Is the name `clickToLoadEntity` what we will definitely use?

**Review by kdzwinel — APPROVED** [review]
Thanks for making those changes! This LGTM 🤘

File: mapping.json
Comment (ladamski) [reply mapping.json]: We discussed `clickToLoadEntity` and I don't think we'll need it; we'll have that discussion in the MV3 CTL TD.

Going to merge here, thanks all for your careful reviews and insights!

File: mapping.json
Comment (jdorweiler) [inline mapping.json:46]: Can you split this into two entries, one for youtube.com and one for youtube-nocookie.com

**Pull request merged**
